### PR TITLE
Introduce cartesian axis breaks

### PR DESCRIPTION
**Pull request by etpinard**
which presents a solution for removing weekends, holidays and non-work hours as requested in https://github.com/plotly/plotly.js/issues/1382 and lays the groundwork from "visually skipped" ranges as in the screenshot in https://github.com/plotly/plotly.js/issues/4210.

----

In brief, this PR adds a `breaks` array attribute to cartesian axes. Each item inside breaks defines an axis break either using a 2-item `bounds` array or a free-length array of `values`. When using `bounds` on date axes, one can set a pattern to generate breaks e.g. over all weekends or non-work hours. When using `values`, one can set the "spread" of each value using `dvalue`. The inclusive/exclusive behaviour at the bounds is determined by the `operation` attributes (as used in filter transforms).

Some examples:

- breaks on linear x and y axes: https://codepen.io/etpinard/pen/eYNErKK?editors=1010
- breaks using `pattern` on date axis: https://codepen.io/etpinard/pen/MWwvGBV?editors=1010
- breaks using `values` on date axis: https://codepen.io/etpinard/pen/PoqKedG

To implement axis `breaks`,

- we first set coordinates that fall within the breaks to `BADNUM` (more info in https://github.com/plotly/plotly.js/commit/baf753aad7cb77d9dff47cf946a77d67ba2aca9e). Please note that this implies that lines remain continuous across a break if there's no (x,y) pt inside the break in question.
- sanitize the breaks present inside the axis range during `ax.setScale` (i.e. when the range is known) and adapt the `l2p` and `p2l` mappings (more info in https://github.com/plotly/plotly.js/commit/259eafa1e805def29df18289a0f0292ffda0cc43)
- Use the sum of the piecewise data-space ranges between the breaks (instead of simply `range[1] - range[0]`) for autorange padding, ticks placement computations and during drag interactions.

cc @archmoj @alexcjohnson @nicolaskruchten @antoinerg

### CODE REVIEW

File: src/plots/cartesian/axis_defaults.js [hunk 1]
```diff
@@ -117,5 +121,68 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
...
+                if(rng[0] < rng[1]) {
+                    if(bnds[0] < rng[0] && bnds[1] > rng[1]) {
+                        itemOut.enabled = false;
+                        return;
+                    }
+                } else if(bnds[0] > rng[0] && bnds[1] < rng[1]) {
+                    itemOut.enabled = false;
+                    return;
+                }
+            }
```
Comment (archmoj) [inline src/plots/cartesian/axis_defaults.js]: What about handling `autorange: 'reversed'` case?
In the following demos I tried to flip the axes which it didn't work.
https://codepen.io/MojtabaSamimi/pen/yLNzLra?editors=0010
https://codepen.io/MojtabaSamimi/pen/bGdoGyG?editors=0010
https://codepen.io/MojtabaSamimi/pen/QWbqWRj?editors=0010

Comment (etpinard) [reply src/plots/cartesian/axis_defaults.js]: Ah right, I forgot to test those. They won't be handled here as this logic has to do with set (i.e. `autorange: false` ranges, but yeah I'll get this fixed. Thanks!

File: src/plots/cartesian/set_convert.js [hunk 2]
```diff
@@ -486,6 +549,51 @@ module.exports = function setConvert(ax, fullLayout) {
...
         }
 
+        // set of "N" disjoint breaks inside the range
+        ax._breaks = [];
+        // length of these breaks in value space
+        ax._lBreaks = 0;
+        // l2p slope (same for all intervals)
+        ax._m2 = 0;
+        // set of l2p offsets (one for each of the (N+1) piecewise intervals)
+        ax._B = [];
```
Comment (archmoj) [inline src/plots/cartesian/set_convert.js]: I am afraid using `_B` here may be too abstract. 
How about `_offBr` or other alternatives?

File: src/plots/cartesian/set_convert.js [hunk 3]
```diff
@@ -486,6 +549,51 @@ module.exports = function setConvert(ax, fullLayout) {
             ax._b = -ax._m * rl0;
         }
 
+        // set of "N" disjoint breaks inside the range
+        ax._breaks = [];
+        // length of these breaks in value space
+        ax._lBreaks = 0;
```
Comment (archmoj) [inline src/plots/cartesian/set_convert.js]: `ax._lenBr` may read better.

File: src/plots/cartesian/set_convert.js (hunk 2 above)
Comment (etpinard) [reply src/plots/cartesian/set_convert.js]: I can change it if you want, but `ax._B` refers to `ax._b` which is used on axes w/o breaks.

Comment (archmoj) [reply src/plots/cartesian/set_convert.js]: Then adding a comment may be enough.

File: src/plots/cartesian/set_convert.js [hunk 4]
```diff
@@ -565,6 +673,141 @@ module.exports = function setConvert(ax, fullLayout) {
...
+                                step = ONEDAY;
+
+                                t = r0 + r0PatternDelta * ONEHOUR -
+                                    r0Date.getUTCMinutes() * ONEMIN -
+                                    r0Date.getUTCSeconds() * ONESEC -
+                                    r0Date.getUTCMilliseconds();
+                                break;
+                        }
+
+                        while(t <= r1) {
```
Comment (archmoj) [inline src/plots/cartesian/set_convert.js]: Could we add a finite test for `t` before entring this loop?
Is there any scenario (e.g. on reversed axes) in which step could be negative?

**Comment by nicolaskruchten** [comment]
Looks great! Do we have any control over the visual aspect of the break? Can we put in a gap or a squiggle or something?

**Comment by etpinard** [comment]
> Do we have any control over the visual aspect of the break? Can we put in a gap or a squiggle or something?

Not at the moment.

https://github.com/plotly/plotly.js/blob/187c93afe4f210ceef20a1e67aa66d36109cfec7/src/plots/cartesian/layout_attributes.js#L346-L375

**Comment by nicolaskruchten** [comment]
OK. I feel like introducing breaks without visible gaps is a bit scary in other contexts than "drop weekends for business metrics" ... what's the LOE on visible gaps?

**Comment by etpinard** [comment]
> what's the LOE on visible gaps?

Probably 1-2 days for gaps w/o squiggle on the axis. I suspect that getting everyone happy with the squiggle look/behaviour will take a few days on its own.

File: src/plots/cartesian/layout_attributes.js [hunk 5]
```diff
@@ -248,6 +248,135 @@ module.exports = {
...
+            role: 'info',
+            items: [
+                {valType: 'any', editType: 'calc'},
+                {valType: 'any', editType: 'calc'}
+            ],
+            editType: 'calc',
+            description: [
+                'Sets the lower and upper bounds of this axis break.',
+                'Can be used with `operation` to determine the behavior at the bounds.',
+                'On *date* axes, it can be used `pattern`.'
```
Comment (nicolaskruchten) [inline src/plots/cartesian/layout_attributes.js]: missing "with" before "pattern" ?

**Comment by nicolaskruchten** [comment]
Zooming way in to the pattern date mock yields some odd tick labels: 

image

File: src/plots/cartesian/layout_attributes.js [hunk 6]
```diff
@@ -248,6 +248,135 @@ module.exports = {
...
+            ].join(' ')
+        },
+        dvalue: {
+            // TODO could become 'any' to add support for 'months', 'years'
+            valType: 'number',
+            role: 'info',
+            editType: 'calc',
+            min: 0,
+            description: [
+                'Sets the spread of each `values` item.',
```
Comment (nicolaskruchten) [inline src/plots/cartesian/layout_attributes.js]: I don't really understand what this means... "spread" is unclear in this context, can you clarify please?

Comment (etpinard) [reply src/plots/cartesian/layout_attributes.js]: An axis break is an continuous interval, so when setting an axis break via a set of values, one needs to also specify the distance on the axis that one value corresponds to.

File: src/plots/cartesian/layout_attributes.js (hunk 5 above)
Comment (etpinard) [reply src/plots/cartesian/layout_attributes.js]: Good eye! Fixed in https://github.com/plotly/plotly.js/pull/4614/commits/76a265e75e7f6f67169710ec292b625ef9b8e172

File: src/plots/cartesian/layout_attributes.js (hunk 6 above)
Comment (nicolaskruchten) [reply src/plots/cartesian/layout_attributes.js]: As discussed, I think "size" is less confusing than "spread" and an inline example would help a lot :)

**Comment by nicolaskruchten** [comment]
To capture what was discussed today: for 1.53 let's activate this feature for date axes only please. We'll need to think about axis gap and bar charts etc before turning this on on linear/log axes I think.

**Comment by etpinard** [comment]
TODO: 

- [ ] coerce breaks only on `type: 'date'` axes, adapt tests and mocks!
- [ ] make sure axis breaks work on axes with `range[0] > range[1]`

see comments over on: https://github.com/plotly/plotly.js/compare/axis-breaks...axis-breaks-fixups

@archmoj will attempt to fix these tomorrow and on Monday the 9th. 

- [ ] better "first tick" algorithm e.g. what to do when the canonical first auto tick falls into a break? See https://github.com/plotly/plotly.js/pull/4614/commits/b5aaf92ce91e7f76101cfebd41ccfa101d747521
- [ ] better "overlapping tick in either side of an axis break fix" algorithm, see comment in https://github.com/plotly/plotly.js/pull/4614/commits/6bec94d3869b8af7a9260142e5127bc95a1662a8

I should be able to look at these when I get back Tuesday the 10th.

**Comment by archmoj** [comment]
There is a console error in this [demo](https://codepen.io/MojtabaSamimi/pen/eYNyBZV?editors=0010).

**Comment by archmoj** [comment]
@etpinard my branch in at https://github.com/plotly/plotly.js/tree/axis-breaks-fixups_ms

File: src/plots/cartesian/layout_attributes.js (hunk 6 above)
Comment (etpinard) [reply src/plots/cartesian/layout_attributes.js]: Done in https://github.com/plotly/plotly.js/pull/4614/commits/7080f900074a59c6af3fd664999d8493e9d2fe84

**Comment by etpinard** [comment]
> There is a console error in this [demo](https://codepen.io/MojtabaSamimi/pen/eYNyBZV?editors=0010).

Looks like this one is just an example of a more general bug. Commenting out the `axis.breaks` and replacing the weekends dates with `NaN`s gives the same console error: https://codepen.io/etpinard/pen/dyoJxKe?editors=0010

@archmoj you've worked on this routine in `bar/plot.js` recently, do you have an idea why this happens?

**Comment by archmoj** [comment]
> > There is a console error in this [demo](https://codepen.io/MojtabaSamimi/pen/eYNyBZV?editors=0010).
> 
> Looks like this one is just an example of a more general bug. Commenting out the `axis.breaks` and replacing the weekends dates with `NaN`s gives the same console error: https://codepen.io/etpinard/pen/dyoJxKe?editors=0010
> 
> @archmoj you've worked on this routine in `bar/plot.js` recently, do you have an idea why this happens?

Something went wrong in [v1.50.0](https://github.com/plotly/plotly.js/releases/tag/v1.50.0) | [demo-console-error](https://codepen.io/MojtabaSamimi/pen/abOEerE?editors=1000).
Possibly a side effect of `Add transition support to bar trace length, width, on-graph text positioning,
marker style and error bars` [#4180, #4186]

**Comment by archmoj** [comment]
It could be related to 

Screenshot from 2020-03-10 15-18-07

in https://github.com/plotly/plotly.js/pull/4180/commits/f49828a05fdf593097083842b31cbc2f8f777c11

**Comment by archmoj** [comment]
> TODO:
> 
> * [ ]  make sure axis breaks work on axes with `range[0] > range[1]`

On a reversed x-axis the hover position sometimes is in the wrong position: [demo](https://codepen.io/MojtabaSamimi/pen/JjdpPNm?editors=1000).

**Comment by archmoj** [comment]
On axes with `autorange: 'reversed'` of this [demo](https://codepen.io/MojtabaSamimi/pen/mdJXbQe?editors=0010) autorange does not fit.

**Comment by archmoj** [comment]
It appears (at the moment) `cartesian` `breaks` are not compatible with `gl2d` traces namely `scattergl`: [demo](https://codepen.io/MojtabaSamimi/pen/OJVQJXg?editors=0010). 
Should we warn the user in that case?

**Comment by etpinard** [comment]
> Should we warn the user in that case?

Fixing the behaviour for `scattergl` should not be hard.

**Comment by etpinard** [comment]
https://github.com/plotly/plotly.js/pull/4614/commits/dcceb7693ed2a68264a295b9dca22c2e2ec65224 fixes the overlapping ticks of https://github.com/plotly/plotly.js/pull/4614#issuecomment-595242555

Updated codepen: https://codepen.io/etpinard/pen/rNVJaex?editors=1010

**Comment by archmoj** [comment]
> > TODO:
> > 
> > * [ ]   make sure axis breaks work on axes with `range[0] > range[1]`
> 
> On a reversed x-axis the hover position sometimes is in the wrong position: [demo](https://codepen.io/MojtabaSamimi/pen/JjdpPNm?editors=1000).

@etpinard please pick https://github.com/plotly/plotly.js/commit/23c566f98b4ab92b463b726eb3132dea279b26f4 to fix hover on reversed ranges.

**Comment by etpinard** [comment]
> @etpinard please pick [23c566f](https://github.com/plotly/plotly.js/commit/23c566f98b4ab92b463b726eb3132dea279b26f4) to fix hover on reversed ranges.

Thanks!! I know I should be doing this, but I'm running out of time: can you try adding a test to lock this down similar to 

https://github.com/plotly/plotly.js/blob/dcceb7693ed2a68264a295b9dca22c2e2ec65224/test/jasmine/tests/hover_label_test.js#L2707-L2773

**Comment by etpinard** [comment]
> > Should we warn the user in that case?
> 
> Fixing the behaviour for `scattergl` should not be hard.

I was wrong. This will actually be very tricky to get right for `scattergl` - most likely `regl-scatter2d`, `regl-line2d` and `regl-error2d` will have to be patched. I doubt I'll have time to do this before Friday.

**Comment by archmoj** [comment]
> > > Should we warn the user in that case?
> > 
> > 
> > Fixing the behaviour for `scattergl` should not be hard.
> 
> I was wrong. This will actually be very tricky to get right for `scattergl` - most likely `regl-scatter2d`, `regl-line2d` and `regl-error2d` will have to be patched. I doubt I'll have time to do this before Friday.

Thanks for the investigation. Could we `visible: false` scattergl on date axes with breaks?

**Comment by alexcjohnson** [comment]
> > Should we warn the user in that case?
> Could we `visible: false` scattergl on date axes with breaks?

Yes, let's leave gl out of this PR. Hide them - and log a warning - in that case. But @archmoj please make an issue to come back to this soon, I can see this being important for some users.

**Comment by etpinard** [comment]
> * better "first tick" algorithm e.g. what to do when the canonical first auto tick falls into a break?

Commit https://github.com/plotly/plotly.js/pull/4614/commits/28c328dfd3f9f31c83b727b610f4329f2b0c6149 made things a bit better. Still room for improvement though.

**Comment by archmoj** [comment]
> > > Should we warn the user in that case?
> > > Could we `visible: false` scattergl on date axes with breaks?
> 
> Yes, let's leave gl out of this PR. Hide them - and log a warning - in that case. But @archmoj please make an issue to come back to this soon, I can see this being important for some users.

Good call. Tracked in #4630.

**Comment by archmoj** [comment]
> It could be related to
> 
> Screenshot from 2020-03-10 15-18-07
> 
> in [f49828a](https://github.com/plotly/plotly.js/commit/f49828a05fdf593097083842b31cbc2f8f777c11)

@etpinard 
Now that `NaN` positions are handled in #4634, 
would you mind merging master to this branch?

**Comment by archmoj** [comment]
Console error when zooming out (+10 times) on this [demo](https://codepen.io/MojtabaSamimi/pen/WNvzQdd?editors=1010).

**Comment by archmoj** [comment]
Another console error when zooming in (+2 times) on this [demo](https://codepen.io/MojtabaSamimi/pen/WNvzQdd?editors=1010).

**Comment by etpinard** [comment]
Nice catch. It's the same problem in both demos. The modebar zoom button calls a relayout with a range falling inside a break. I'm not exactly sure what to do here.

**Comment by archmoj** [comment]
Also to note by panning left and right (n times) the x-axis scale would change (which I assume it shouldn't) until the line would be drawn vertically where get an error.

**Comment by etpinard** [comment]
> Also to note by panning left and right (n times) the x-axis scale would change (which I assume it shouldn't) until the line would be drawn vertically where get an error.

Hmm. That shouldn't happen. The logic here must be off:

https://github.com/plotly/plotly.js/blob/28c328dfd3f9f31c83b727b610f4329f2b0c6149/src/plots/cartesian/dragbox.js#L1017-L1020

**Comment by archmoj** [comment]
Thanks @etpinard the following can help:
```js
var p0 = 0;
var p1 = axi._length;
var d0 = axi.p2l(p0 + pix) - axi.p2l(p0);
var d1 = axi.p2l(p1 + pix) - axi.p2l(p1);
var delta = (d0 + d1) / 2;

axi.range = [
    axi.l2r(axi._rl[0] - delta),
    axi.l2r(axi._rl[1] - delta)
];
```

**Comment by etpinard** [comment]
> > * better "first tick" algorithm e.g. what to do when the canonical first auto tick falls into a break?
> 
> Commit [28c328d](https://github.com/plotly/plotly.js/commit/28c328dfd3f9f31c83b727b610f4329f2b0c6149) made things a bit better. Still room for improvement though.

Improved once more in https://github.com/plotly/plotly.js/pull/4614/commits/1b55b42f5ce3c6986774c3866d70630a72d6721a

-------

Pan fixes from @archmoj are in https://github.com/plotly/plotly.js/pull/4614/commits/40d57fa6c4e29a5fb7d10cf528ef9975f08792f5

--------



> > > Should we warn the user in that case?
> > > Could we `visible: false` scattergl on date axes with breaks?
> 
> Yes, let's leave gl out of this PR. Hide them - and log a warning - in that case. But @archmoj please make an issue to come back to this soon, I can see this being important for some users.

Done in https://github.com/plotly/plotly.js/pull/4614/commits/49b40534107c1d5cf91b9765e38e2a79064e776c

**Comment by archmoj** [comment]
Great PR! 
:medal_military: :medal_military: :1st_place_medal: 

Thanks very much @etpinard for tackling all this.
:dancer: :man_dancing:

**Comment by etpinard** [comment]
Ok, merging to get a first round of QA rolling (I suspect a few bug reports will come out of it).

Update examples:

breaks using pattern on date axis: https://codepen.io/etpinard/pen/PoqRNoL
breaks using values on date axis: https://codepen.io/etpinard/pen/OJVvNJG

**Comment by nicolaskruchten** [comment]
🏆 thanks @etpinard 🙇

**Comment by kenahoo** [comment]
Is this functionality part of any release?  Or is it still experimental?  I'm having a little trouble finding how to do it in the docs.

**Comment by archmoj** [comment]
> Is this functionality part of any release? Or is it still experimental? I'm having a little trouble finding how to do it in the docs.

@kenahoo yes it is available since `v1.53.0`
You could find javascript documentation at https://plotly.com/javascript/reference/layout/xaxis/#layout-xaxis-rangebreaks.
The documentation for Python at the bottom of this page https://plotly.com/python/time-series/ could also be of interest.

**Pull request merged**